### PR TITLE
Migration to PJ and various memleak fixes

### PR DESCRIPTION
**Pull request by XVilka**
DO NOT SQUASH

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/rizinbook) with the relevant information (if needed)

**Detailed description**

Sync to the latest radare2, but not everything is imported - only minor changes.

### CODE REVIEW

**Review by wargio — COMMENTED** [review]
can't we use the PJ pointer to check if the output is JSON or not, instead of the `int mode` variable?

**Comment by Maijin** [comment]
Can you add a test for `arS`?

**Comment by officialcjunior** [comment]
> Can you add a test for `arS`?

How exactly should a file be opened to test that?

**Comment by XVilka** [comment]
@wargio we can, but it's out of the scope of this PR. Can be done later.

**Comment by XVilka** [comment]
Ugnore the `jni.h` failed test, it will be green once https://github.com/rizinorg/rizin-testbins/pull/4 is merged.

File: librz/debug/dsession.c [hunk 1]
```diff
@@ -576,6 +580,7 @@ static bool deserialize_checkpoints_cb(void *user, const char *cnum, const char
 	// Extract RzRegArena's from "registers"
 	const RJson *regs_json = rz_json_get (chkpt_json, "registers");
 	if (!regs_json || regs_json->type != RZ_JSON_ARRAY) {
+		free (json_str);
```
Comment (ret2libc) [inline librz/debug/dsession.c:583]: there should be also `rz_json_free (chkpt_json);` after this line AFAIK

File: test/unit/test_str.c [hunk 2]
```diff
@@ -558,14 +558,20 @@ bool test_rz_str_encoded_json(void) {
 
 	const char *array = rz_str_encoded_json (invalidJsonString, len, PJ_ENCODING_STR_ARRAY);
 	mu_assert_streq (array, "084,104,105,115,032,105,115,032,109,121,032,226,032,115,097,109,112,108,101,032,194,169,032,115,116,114,105,110,103,010", "string as array of uchar");
```
Comment (ret2libc) [inline test/unit/test_str.c]: use `mu_assert_streq_free` instead of `mu_assert_streq` + `free`.

File: librz/cons/line.c [hunk 3]
```diff
@@ -41,6 +41,7 @@ RZ_API void rz_line_free(void) {
 	// XXX: prompt out of the heap?
 	free ((void *)I.prompt);
 	I.prompt = NULL;
+	//rz_list_free (I.kill_ring);
```
Comment (ret2libc) [inline librz/cons/line.c:44]: why this commented line?

Comment (ret2libc) [reply librz/cons/line.c:44]: I think it is right to uncomment it.

Comment (XVilka) [reply librz/cons/line.c:44]: It causes double free segfault then.

Comment (ret2libc) [reply librz/cons/line.c:44]: Why? Does it happen in radare2 as well? Maybe it should be `rz_list_free`+`I.kill_ring=NULL`? (though this is not *the* solution)

Comment (XVilka) [reply librz/cons/line.c:44]: It doesn't happen in radare2:  https://github.com/radareorg/radare2/commit/06b735441d299e90695235495982354bef1fcc41 - note the green status.

Comment (XVilka) [reply librz/cons/line.c:44]: I can just remove this line if you want.

Comment (ret2libc) [reply librz/cons/line.c:44]: Adding this free is right in theory... Could you try uncommenting it so I can see where exactly the double free happens?

Comment (XVilka) [reply librz/cons/line.c:44]: What about merging this first, then try? It's grown quite big.

**Pull request merged**